TestingApp/Extensions/Configuration.Extensions.swift

```diff
@@ -0,0 +1,159 @@
+        switch rawValue {
+        case "beta":
+            self = .beta
+        case "us":
+            self = .usa
+        case "eu":
+            self = .europe
+        default:
+            guard let url = URL(string: rawValue) else { return nil }
+            self = .custom(url)
+        }
+    }
+}
+
+extension Configuration: Codable {
+
+    enum CodingKeys: String, CodingKey {
+        case authorizationMethod, environment, site, visitorContext
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self = try .init(
+            authorizationMethod: container.decode(Configuration.AuthorizationMethod.self, forKey: .authorizationMethod),
+            environment: container.decode(Environment.self, forKey: .environment),
+            site: container.decode(String.self, forKey: .site),
+            visitorContext: container.decodeIfPresent(VisitorContext.self, forKey: .visitorContext)
+        )
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(authorizationMethod, forKey: .authorizationMethod)
+        try container.encode(environment, forKey: .environment)
+        try container.encode(site, forKey: .site)
+        try container.encode(visitorContext, forKey: .visitorContext)
+    }
+}
+
+extension Configuration.AuthorizationMethod: Codable {
+
+    enum CodingKeys: String, CodingKey {
+        case id, secret, appToken
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        if container.contains(.appToken) {
+            self = try .appToken(container.decode(String.self, forKey: .appToken))
+        } else {
+            self = try .siteApiKey(
+                id: container.decode(String.self, forKey: .id),
+                secret: container.decode(String.self, forKey: .secret)
+            )
+        }
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        switch self {
+        case .appToken(let appToken):
+            try container.encode(appToken, forKey: .appToken)
+        case let .siteApiKey(id, secret):
+            try container.encode(id, forKey: .id)
+            try container.encode(secret, forKey: .secret)
+        }
+    }
+}
+
+extension Environment: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let rawValue = try container.decode(String.self)
+        switch rawValue {
+        case "beta":
+            self = .beta
+        case "us":
+            self = .usa
+        case "eu":
+            self = .europe
+        default:
+            guard let custom = URL(string: rawValue) else {
+                throw NSError(domain: "not-valid-env", code: -33)
```

Comment:
Where does this code comes from? the -33

Reply:
this error code will be available for internal usage only and only in Testing app. I changed code to 0 - ac2f9517dd53b07be67dce8259cee91caa1eeb83